src/lib/LQCALC/LQCALC.h

```diff
@@ -1,23 +1,80 @@
 #pragma once
 
 #include <Arduino.h>
-#include <stdio.h>
 #include "../../src/targets.h"
 
+template <uint8_t N>
 class LQCALC
 {
 public:
-    void init(uint8_t depth);
-    void init();
-    void ICACHE_RAM_ATTR add();
-    void ICACHE_RAM_ATTR inc();
-    uint8_t ICACHE_RAM_ATTR getLQ();
-    void ICACHE_RAM_ATTR reset();
+    LQCALC(void)
+    {
+        reset();
+    }
+
+    /* Set the bit for the current period to true and update the running LQ */
+    void ICACHE_RAM_ATTR add()
+    {
+        LQArray[LQbyte] |= LQmask;
+        LQ += 1;
+    }
+
+    /* Start a new period */
+    void ICACHE_RAM_ATTR inc()
+    {
+        // Increment the counter by shifting one bit higher
+        // If we've shifted out all the bits, move to next idx
+        LQmask = LQmask << 1;
+        if (LQmask == 0)
+        {
+            LQmask = (1 << 0);
+            LQbyte += 1;
+        }
+
+        // At idx N / 32 and bit N % 32, wrap back to idx=0, bit=0
+        if ((LQbyte == (N / 32)) && (LQmask & (1 << (N % 32))))
+        {
+            LQbyte = 0;
+            LQmask = (1 << 0);
+        }
+
+        if ((LQArray[LQbyte] & LQmask) != 0)
+        {
+            LQArray[LQbyte] &= ~LQmask;
+            LQ -= 1;
```

Comment:
LQ -= 1 should be outside of the if condition???

Currently if you have 2 missed packets that are N apart then LQ will not be decrease on the seconded missed packet.

Reply:
No, it should not be outside the if condition. You can only subtract one LQ if you added one LQ when that packet was received. Therefore you only subtract the LQ when the bit you're clearing is set.

Your second statement I don't follow your logic. If the array is full of 1s (N=100, LQ=100) the following effective code executes and reports 99 as expected. 
```
  l.inc();
  // <-- no add, missed packet
  // vv N adds
  for (int i=0; i<N; ++i)
  {
    l.inc();
    l.add();
  }
  l.inc();
  // <-- no add, missed packet
  // vv reports 99
  printf("Pass receives=%d\n", l.getLQ());
```
The code only adds one when `add()` is called, therefore it only subtracts one when the packet tally being shifted out was one. The bit always needs to be & masked, but because it only really needs to be cleared if it was set, moving that & mask operation inside the if is a microptimization.